### ISSUE TITLE
When entering a password in the Mirth Server Manager, some characters are escaped before being saved.

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#1408]
If I use a database password with special characters and enter it in Mirth Server Manager, certain special characters are escaped with a backslash before getting stored into the file sqlserver2005-SqlMapConfig.properties. Consequently, the password does not work. Characters that have this issue are:

backslash \ exclamation mark ! hash # and full colon :

two workarounds:
1. don't use those characters in a password
2. add passwords directly to sqlserver2005-SqlMapConfig.properties instead of using the Mirth Server Manager

Imported Issue. Original Details:
Jira Issue Key: MIRTH-1409
Reporter: colinsj
Created: 2010-04-29T08:38:03.000-0700

**Comment by rbeckman-nextgen** [comment]
I verified that the Properties load and store is working correctly: it escapes the special characters on store and removes the slashes on load. Based on this, it looks like the issue is when the password properties is being resolved when embedded in the SqlMapConfig.xml file.

Imported Comment. Original Details:
Author: geraldb
Created: 2010-05-11T10:07:46.000-0700

**Comment by rbeckman-nextgen** [comment]
Looks like a bug with the version of ibatis we are using (2.3.4).

Imported Comment. Original Details:
Author: jacobb
Created: 2010-05-19T14:47:28.000-0700

**Comment by rbeckman-nextgen** [comment]
I'm not able to reproduce this... using Windows 7, SQL Server 2005 Express, MC 1.8.2, Java 1.6.0_45. It just seems to work, no matter what configuration I throw at it.

Imported Comment. Original Details:
Author: narupley
Created: 2013-06-13T17:22:15.000-0700

**Comment by rbeckman-nextgen** [comment]
Crazy. We don't really need this fixed at this point, and we're migrating to a more up to date Mirth this year, but I tried this out again to sanity check my original report. I hope this comment form preserves line feeds. Here we go:

I set the password on SQL Server to "a\!90" 

Then, I use the Mirth Connect Server Manager (Database Tab) to set that password. Here are the field values:
   Type: sqlserver2005
   Driver: net.sourceforge.jtds.jdbc.Driver
   URL: jdbc:jtds:sqlserver://integrationsqltest:1433/MIRTH_TESTING;instance=Mirth
   Username: Mirth
   Password: a\!90

I click ok, and then open file sqlserver2005-SqlMapConfig.properties. On the second line is
    password=a\\\!90

I restart the Mirth service, but Mirth doesn't work until I manually change that line to contain
    password=a\!90

Mirth Connect Administrator 1.8.2.4472 with Java 1.6.0_23 running on
Windows 2003, SP2

SQL Server 2005 (9.0.3042) also running on Windows 2003, SP2 (separate instance)

Imported Comment. Original Details:
Author: colinsj
Created: 2013-06-17T16:08:52.000-0700

**Comment by rbeckman-nextgen** [comment]
Foiled. Your webform stripped out some, but not all, of my backslashes.

Imported Comment. Original Details:
Author: colinsj
Created: 2013-06-17T16:10:23.000-0700

**Comment by rbeckman-nextgen** [comment]
Colin, you can try using the code macro \` around your text:

`
Example
\
\\
\\\
\\\\
`

Imported Comment. Original Details:
Author: jacobb
Created: 2013-06-17T16:28:38.000-0700

**Comment by rbeckman-nextgen** [comment]
I hope I get this right. I assume the macro start tag and end tag are identical since you didn't specify otherwise. And seriously, we really don't need this fixed, I just felt compelled to re-test given your results. Here is my previous post, surrounded by the code macro (hopefully correctly) with the two missing backslashes restored.
`
Crazy. We don't really need this fixed at this point, and we're migrating to a more up to date Mirth this year, but I tried this out again to sanity check my original report. I hope this comment form preserves line feeds. Here we go:

I set the password on SQL Server to "a!90" 

Then, I use the Mirth Connect Server Manager (Database Tab) to set that password. Here are the field values:
 Type: sqlserver2005
 Driver: net.sourceforge.jtds.jdbc.Driver
 URL: jdbc:jtds:sqlserver://integrationsqltest:1433/MIRTH_TESTING;instance=Mirth
 Username: Mirth
 Password: a\!90

I click ok, and then open file sqlserver2005-SqlMapConfig.properties. On the second line is
 password=a\\\!90

I restart the Mirth service, but Mirth doesn't work until I manually change that line to contain
 password=a\!90

Mirth Connect Administrator 1.8.2.4472 with Java 1.6.0_23 running on
 Windows 2003, SP2

SQL Server 2005 (9.0.3042) also running on Windows 2003, SP2 (separate instance)
`

Imported Comment. Original Details:
Author: colinsj
Created: 2013-06-17T16:41:44.000-0700

**Comment by rbeckman-nextgen** [comment]
Oh, there was a third spot where the backslash was stripped out. It should have read: `I set the password on SQL Server to "a\!90"`

Imported Comment. Original Details:
Author: colinsj
Created: 2013-06-17T16:43:08.000-0700

**Comment by rbeckman-nextgen** [comment]
So the actual password was "a!90" but you typed the characters

`a\!90`

into the server manager? That seems like the expected behaviour then. If you only typed "a!90" into the server manager, the resulting SqlMapConfig.properties file would contain:

`a\!90`

meaning that the actual password (post-escaping) is "a!90".

Imported Comment. Original Details:
Author: narupley
Created: 2013-06-17T17:09:31.000-0700

**Comment by rbeckman-nextgen** [comment]
`So is the problem then that SQL Server / ibatis was requiring the ! to be escaped with a \?`

Imported Comment. Original Details:
Author: jacobb
Created: 2013-06-17T17:12:24.000-0700

**Comment by rbeckman-nextgen** [comment]
Just tested it on Windows XP and SQL Server 2005; no change. I set the SQL Server password to `a\!90`, typed in `a\!90` in the server manager, and everything worked without a hitch.

Imported Comment. Original Details:
Author: narupley
Created: 2013-06-17T17:15:49.000-0700

**Comment by rbeckman-nextgen** [comment]
Try setting the sql server password to a!90.

I think the issue is that that wasn't going through ibatis properly...

Imported Comment. Original Details:
Author: jacobb
Created: 2013-06-17T17:17:32.000-0700

**Comment by rbeckman-nextgen** [comment]
I tried "a!90", no change... everything works as expected.

Imported Comment. Original Details:
Author: narupley
Created: 2013-06-17T17:19:47.000-0700

**Comment by rbeckman-nextgen** [comment]
Try a!90 in 2.2.

Imported Comment. Original Details:
Author: jacobb
Created: 2013-06-17T17:24:08.000-0700

**Comment by rbeckman-nextgen** [comment]
Both "a!90" and `a\!90` work in 2.2.2 as well..

Imported Comment. Original Details:
Author: narupley
Created: 2013-06-17T17:44:28.000-0700

**Comment by rbeckman-nextgen** [comment]
Oh boy, the confusion that commences when characters go missing.
`
The actual password was: a\!90
That's five characters
a (eh, I'm Canadian)
\ (backslash)
! (exclamation mark)
9 (nine)
0 (zero)

After inputting it into the Server Manager, it becomes seven characters long in sqlserver2005-SqlMapConfig.properties.
a (eh, like most Western Canadians, I pronounce "about" with a dipthong ah-b(ah->oh)t)
\ (backslash)
\ (backslash)
\ (backslash)
! (exclamation mark)
9 (nine)
0 (zero)
`
We are migrating to 2.2.2 this year. All will be well.

Imported Comment. Original Details:
Author: colinsj
Created: 2013-06-17T20:28:01.000-0700

**Comment by rbeckman-nextgen** [comment]
In 1.8.2 when you type those five characters into the server manager, the actual value that shows up in the properties file is:

`a\\\!90`

However in 2.2.2, the actual value that shows up is:

`a\\!90`

This is because we changed how we store those properties. In 1.8.2, we set them using [Properties.store](http://docs.oracle.com/javase/6/docs/api/java/util/Properties.html#store%28java.io.Writer,%20java.lang.String%29):

{quote}
For the key, all space characters are written with a preceding \ character. For the element, leading space characters, but not embedded or trailing space characters, are written with a preceding \ character. The key and element characters #, !, =, and : are written with a preceding backslash to ensure that they are properly loaded. 
{quote}

In 2.2.2, we use commons-configuration instead, meaning that only backslashes are escaped and nothing else.

Also, we changed how we _load_ those properties between the versions as well. In 1.8.2 we left it up to iBATIS to parse the properties file, whereas in 2.2.2 we load them manually via commons-configuration and then pass them into the SqlMapClientBuilder directly.

However in both versions, it correctly loads the properties on server startup, and successfully connects to the database. In fact, it doesn't even **matter** (in either version) whether you have `a\\!90` or `a\\\!90`. The properties will still be correctly read either way. In other words, both values are correct.

Now this is hypothetical, but say the actual database password was "a!90" and you _thought_ it was `a\!90`. So you typed those five characters into the server manager, and the resulting password value in the properties file was `a\\\!90`. Then, thinking that it's incorrect, you changed the value in the properties file to `a\!90`, and then voilà, the server starts up. That chain of events would _make_ _sense_, because a value in the properties file of `a\!90` would actually evaluate to "a!90", which is the correct password.

Anyway, I have yet to reproduce the behaviour where it does **not** successfully connect to the database on startup, in either 1.8.2 or 2.2.2.

Imported Comment. Original Details:
Author: narupley
Created: 2013-06-18T09:26:56.000-0700
